test/StreamJsonRpc.Tests/JsonContractResolverTests.cs

```diff
@@ -0,0 +1,108 @@
+using Microsoft.VisualStudio.Threading;
+using Nerdbank.Streams;
+using Newtonsoft.Json.Serialization;
+using StreamJsonRpc;
+using Xunit;
+using Xunit.Abstractions;
+
+public partial class JsonContractResolverTest : TestBase
+{
+    protected readonly Server server = new Server();
+    protected readonly JsonRpc serverRpc;
+    protected readonly JsonRpc clientRpc;
+    protected readonly IServer client;
+
+    private const string ExceptionMessage = "Some exception";
+
+    public JsonContractResolverTest(ITestOutputHelper logger)
+        : base(logger)
+    {
+        var pipes = FullDuplexStream.CreatePipePair();
+
+        this.client = JsonRpc.Attach<IServer>(new LengthHeaderMessageHandler(pipes.Item1, this.CreateFormatter()));
+        this.clientRpc = ((IJsonRpcClientProxy)this.client).JsonRpc;
+
+        this.serverRpc = new JsonRpc(new LengthHeaderMessageHandler(pipes.Item2, this.CreateFormatter()));
+        this.serverRpc.AddLocalRpcTarget(this.server);
+
+        this.serverRpc.TraceSource = new TraceSource("Server", SourceLevels.Verbose);
+        this.clientRpc.TraceSource = new TraceSource("Client", SourceLevels.Verbose);
+
+        this.serverRpc.TraceSource.Listeners.Add(new XunitTraceListener(this.Logger));
+        this.clientRpc.TraceSource.Listeners.Add(new XunitTraceListener(this.Logger));
+
+        this.serverRpc.StartListening();
+    }
+
+    protected interface IServer : IDisposable
+    {
+        Task PushCompleteAndReturn(IObserver<int> observer);
+    }
+
+    [Fact]
+    public async Task PushCompleteAndReturn()
+    {
+        var observer = new MockObserver<int>();
+        await Task.Run(() => this.client.PushCompleteAndReturn(observer)).WithCancellation(this.TimeoutToken);
+        await observer.Completion.WithCancellation(this.TimeoutToken);
+        ImmutableList<int> result = await observer.Completion;
+    }
+
+    protected IJsonRpcMessageFormatter CreateFormatter()
+    {
+        var formatter = new JsonMessageFormatter();
+        formatter.JsonSerializer.ContractResolver = new CamelCasePropertyNamesContractResolver();
+
+        return formatter;
+    }
+
+    protected class Server : IServer
+    {
+        public Task PushCompleteAndReturn(IObserver<int> observer)
+        {
+            for (int i = 1; i <= 3; i++)
+            {
+                observer.OnNext(i);
+            }
+
+            observer.OnCompleted();
+            return Task.CompletedTask;
+        }
+
+        void IDisposable.Dispose()
+        {
+        }
+    }
+
+    protected class MockObserver<T> : IObserver<T>
+    {
+        private readonly TaskCompletionSource<ImmutableList<T>> completed = new TaskCompletionSource<ImmutableList<T>>();
+
+        internal event EventHandler<T>? Next;
+
+        [System.Runtime.Serialization.IgnoreDataMember]
+        internal ImmutableList<T> ReceivedValues { get; private set; } = ImmutableList<T>.Empty;
+
+        [System.Runtime.Serialization.IgnoreDataMember]
```

Comment:
I'm puzzled as to why these `IgnoreDataMember` attributes are necessary. Surely the stock implementations of this interface that we use already have these applied. Were they necessary before your original object marshaling PR?

Reply:
These are copied from the original `IObserver` tests, I removed them now. I don't think they have any effect on the test.